### PR TITLE
Implement singly linked list

### PR DESCRIPTION
**Pull request by rouille**
### Purpose
Implement singly linked list

### What the code is doing
Two classes are created, i) the `Node` class representing an element (data and reference to next node) of the singly linked list and ii) the `SinglyLinkedLast` creating the linked list and adding/removing elements to it.

### Testing
New unit tests

### Where to look
A new **data_structure** enclosing the `singly_linked_list` module is created.

### Usage Example/Visuals
N/A

### Time estimate
15min

### CODE REVIEW

File: interview/data_structure/singly_linked_list.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+        :param list data: initial element(s) to create linked list.
+        """
+        if data is not None:
+            nodes = data.copy()
+            node = Node(nodes.pop(0))
+            self.head = node
+            for n in nodes:
+                node.next = Node(n)
+                node = node.next
```
Comment (BainanXia) [inline interview/data_structure/singly_linked_list.py]: Maybe we could do
```python
if data:
    for nodeval in data:
        self.add(nodeval)
```

File: interview/data_structure/singly_linked_list.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    def add(self, data, position=None):
```
Comment (BainanXia) [inline interview/data_structure/singly_linked_list.py]: Not sure whether it is more intuitive to break `add` into `insert` and `append`.

File: interview/data_structure/singly_linked_list.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        :param int position: position of the node to remove.
+        :raises ValueError: if position is incorrect.
+        """
+        if self.head is None:
+            raise ValueError("Linked list is empty")
+        elif position == 0:
+            self.head = self.head.next
+        else:
+            for i, n in enumerate(self):
+                if n is not None and n.next is not None and i == (position - 1):
```
Comment (BainanXia) [inline interview/data_structure/singly_linked_list.py]: I think `n is not None` is redundant here given we are iterating via a `for` loop.

**Comment by kasparm** [comment]
Don't have much time but my friend ChatGPT would suggest the following :)

The code appears to be a Python implementation of a singly linked list. Overall, the code looks good and follows good coding practices. Here are a few comments:

1. In the SinglyLinkedList.create() method, instead of using the copy() method on the input data list, it would be better to just iterate over the input data list and create new nodes for each element.

2. The SinglyLinkedList.add() method could be made more efficient by keeping track of the previous node while iterating over the list to find the position where the new node should be inserted.

3. In the SinglyLinkedList.remove() method, the first if statement is unnecessary since the check for an empty linked list can be done in the second elif statement.

4. The __repr__() method could be improved to remove the need for the "None" at the end of the linked list representation by checking if the node is the last node in the list.

5. The __iter__() method could be made more efficient by using a while loop instead of a for loop.

Here's an updated version of the code with these changes:
```
class Node:
    """Single element of linked list. Each node has two different fields, *'data'*
    containing the value to be stored in the node and *'next'* containing a reference
    to the next element in the list.

    :param str/int/float data: value of the node.
    """

    def __init__(self, data):
        self.data = data
        self.next = None

    def __repr__(self):
        return str(self.data)


class SinglyLinkedList:
    """Linked list as a collection of nodes. First element of the linked list is called
    '*head*'.
    """

    def __init__(self):
        self.head = None

    def create(self, data=None):
        """Create linked list with some elements.

        :param list data: initial element(s) to create linked list.
        """
        if data is not None:
            node = Node(data[0])
            self.head = node
            for n in data[1:]:
                node.next = Node(n)
                node = node.next

    def add(self, data, position=None):
        """Add node to the linked list

        :param str/int/float data: value of the node to be inserted
        :param int position: position where the node will be inserted. If set to None,
            the node will be added at the end of the linked list.
        :raises ValueError: if ``position`` is incorrect.
        """
        node = Node(data)
        if position is None:
            if self.head is None:
                self.head = node
            else:
                prev = None
                for n in self:
                    prev = n
                prev.next = node
        elif position == 0:
            node.next = self.head
            self.head = node
        else:
            prev = None
            for i, n in enumerate(self):
                if i == position:
                    raise ValueError(f"position is out of range. Linked list has {i} elements")
                prev = n
            node.next = prev.next
            prev.next = node

    def remove(self, position):
        """Remove node in the linked list.

        :param int position: position of the node to remove.
        :raises ValueError: if position is incorrect.
        """
        if self.head is None:
            raise ValueError("Linked list is empty")
        elif position == 0:
            self.head = self.head.next
        else:
            prev = None
            for i, n in enumerate(self):
                if i
```

File: interview/data_structure/singly_linked_list.py (hunk 2 above)
Comment (rouille) [reply interview/data_structure/singly_linked_list.py]: I like it

File: interview/data_structure/singly_linked_list.py (hunk 3 above)
Comment (rouille) [reply interview/data_structure/singly_linked_list.py]: Indeed

File: interview/data_structure/singly_linked_list.py (hunk 1 above)
Comment (rouille) [reply interview/data_structure/singly_linked_list.py]: Done

File: interview/data_structure/singly_linked_list.py (hunk 2 above)
Comment (rouille) [reply interview/data_structure/singly_linked_list.py]: Done

File: interview/data_structure/singly_linked_list.py (hunk 3 above)
Comment (rouille) [reply interview/data_structure/singly_linked_list.py]: Done

**Review by BainanXia — APPROVED** [review]
Nice.

**Pull request merged**